### ISSUE TITLE
AIRFLOW__CELERY__BROKER_URL not supporting multiple sentinel url, instead throwing error

### ISSUE THREAD

**Issue by amitb2050** [apache/airflow#28655]
### Apache Airflow version

2.5.0

### What happened

I am trying to set up Airflow Celery to use sentinel. However, we have multiple sentinel instances up and running. So referring to the examples on the internet, I tried to pass multiple sentinel URL as below in ENV VAR.

`AIRFLOW__CELERY__BROKER_URL='sentinel://redis_sentinel1:26379;sentinel://redis_sentinel2:26379'` (tried with/without quote)

After executing, the worker keeps throwing errors as below:

> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "/usr/local/lib/python3.8/urllib/parse.py", line 177, in port
>     raise ValueError(message) from None
> **ValueError: Port could not be cast to integer value as '26379;sentinel:'**
> 

From the error, it looks like the parser is NOT separating sentinel URLs with a semicolon separator. Is there any workaround for it?

### What you think should happen instead

Using a semicolon separator, sentinel URLs should be separated, and the Redis connection pool should make use of all the sentinel URLs.

### How to reproduce

Try to pass multiple sentinel URLs to `AIRFLOW__CELERY__BROKER_URL` environment variables and then try to bring up the airflow workers

### Operating System

ubuntu

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Basically a duplicate of #28010 - seems that sentinel does not play well with Airflow and the reason for it is unknown. Would be great if someone who is more familiar with it could take a look. 

I do not think there is a workaround currently but if sufficient number of those who can try and test it could take a loo at this together, maybe they can figure it out, so better discuss it in one issue.

Closing as duplicate of #28010

**Comment by amitb2050** [comment]
Hi, @potiuk Thanks for your response. I went through issue #28010 which concerns why Sentinel doesn't work with TLS setup. However, this issue is related to the error we get when passing multiple sentinel URLs.

Expectation: Multiple sentinel URLs should be accepted per Celery [documentation](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/redis.html) .

**Comment by potiuk** [comment]
> Hi, @potiuk Thanks for your response. I went through issue https://github.com/apache/airflow/issues/28010 which concerns why Sentinel doesn't work with TLS setup. However, this issue is related to the error we get when passing multiple sentinel URLs.

> Expectation: Multiple sentinel URLs should be accepted per Celery [documentation](https://docs.celeryq.dev/en/stable/getting-started/backends-and-brokers/redis.html) .

Yes - and as you can find in https://github.com/apache/airflow/issues/28010 - they seem to be accepted for the user there. Just SSL configuration seems to be not working. In any way this means that Celery Sentinel (for whatever reason) is not working wiht current Celer approach of Ariflow. I would appreciate if you and other users of Celery could help to investigate why and possibly helped to solve it together. This is the fastest way you can help and give back to the community for the free software you get - you seem to be technical enough and have celery sentinel to test it with and incentive to get it working. PRs fixing that are most welcome. And it is best to keep it in a single open issue (linked to that one so that one is not lost and you can see it refers to the same case). 

In short: 
* Yes, it seems there are issues with sentinel
* It's better to consolidate it in one issue
* Since this is a niche case, that might become impotant, as a community we'd appreciate those more technology savvy users who want to use sentinel to help with diagnosing and solving it. 

Any help there is much appreciated.

**Comment by potiuk** [comment]
BTW. This is how OSS development works. Airflow has > 2300 contributors - and vast majority of them are users fixing the problems they deeply care about. You can become one of them if you help to fix/diagnose it and your PR fixing it gets merged.  This is very diferent from commercial software.

**Comment by amitb2050** [comment]
Thanks, @potiuk  for your quick response. Let me study the existing functionality of Airflow CeleryExecutor and see if I can make any modifications to get this working.

**Comment by dintorf** [comment]
I know this issue is closed, however I believe I have gotten Airflow working with Redis Sentinel. Here are the configuration details:

```bash
# airflow config for celery using redis sentinel

# note the broker URL password is the redis password, and the redis DB is 0
AIRFLOW__CELERY__BROKER_URL=sentinel://:str0ng_passw0rd@redis-sentinel1:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel2:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel3:26379/0
AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME=mymaster
# note this is the password for sentinel
AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS={"password":"str0ng_passw0rd"}
AIRFLOW__CELERY__CELERY_CONFIG_OPTIONS=config.celery_config.CELERY_CONFIG
```

The broker transport options are put into celery config using `conf.getsection("celery_broker_transport_options")` which does not load json strings as dicts. Celery expects the `sentinel_kwargs` to be a dict, so I've added this python file to update the broker transport options:

```python
# /usr/local/airflow/config/celery_config.py
import json

from airflow.config_templates.default_celery import DEFAULT_CELERY_CONFIG

CELERY_CONFIG = DEFAULT_CELERY_CONFIG
if "sentinel_kwargs" in CELERY_CONFIG.get("broker_transport_options", {}):
    CELERY_CONFIG["broker_transport_options"]["sentinel_kwargs"] = json.loads(
        CELERY_CONFIG["broker_transport_options"]["sentinel_kwargs"]
    )
```

I've only tested this locally. However, I believe this would work anywhere.

**Comment by potiuk** [comment]
@dintorf - Might be a good idea to contribute that as an example / special case description here: https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/executor/celery.html ?

Just click "Suggest a change on this page"

**Comment by dy957** [comment]
> I know this issue is closed, however I believe I have gotten Airflow working with Redis Sentinel. Here are the configuration details:
> 
> 
> 
> ```bash
> 
> # airflow config for celery using redis sentinel
> 
> 
> 
> # note the broker URL password is the redis password, and the redis DB is 0
> 
> AIRFLOW__CELERY__BROKER_URL=sentinel://:str0ng_passw0rd@redis-sentinel1:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel2:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel3:26379/0
> 
> AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME=mymaster
> 
> # note this is the password for sentinel
> 
> AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS={"password":"str0ng_passw0rd"}
> 
> AIRFLOW__CELERY__CELERY_CONFIG_OPTIONS=config.celery_config.CELERY_CONFIG
> 
> ```
> 
> 
> 
> The broker transport options are put into celery config using `conf.getsection("celery_broker_transport_options")` which does not load json strings as dicts. Celery expects the `sentinel_kwargs` to be a dict, so I've added this python file to update the broker transport options:
> 
> 
> 
> ```python
> 
> # /usr/local/airflow/config/celery_config.py
> 
> import json
> 
> 
> 
> from airflow.config_templates.default_celery import DEFAULT_CELERY_CONFIG
> 
> 
> 
> CELERY_CONFIG = DEFAULT_CELERY_CONFIG
> 
> if "sentinel_kwargs" in CELERY_CONFIG.get("broker_transport_options", {}):
> 
>     CELERY_CONFIG["broker_transport_options"]["sentinel_kwargs"] = json.loads(
> 
>         CELERY_CONFIG["broker_transport_options"]["sentinel_kwargs"]
> 
>     )
> 
> ```
> 
> 
> 
> I've only tested this locally. However, I believe this would work anywhere.

It works for me, thanks

**Comment by ericxiao251** [comment]
Hey folks, how did you get this to work? I am using the Airflow helm charts and I'm not sure where to put things.

```
AIRFLOW__CELERY__BROKER_URL=sentinel://:str0ng_passw0rd@redis-sentinel1:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel2:26379/0;sentinel://:str0ng_passw0rd@redis-sentinel3:26379/0
AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME=mymaster
# note this is the password for sentinel
AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS={"password":"str0ng_passw0rd"}
AIRFLOW__CELERY__CELERY_CONFIG_OPTIONS=config.celery_config.CELERY_CONFIG
```

Are these in `airflow.cfg` or just `ENV` variables?

**Comment by dintorf** [comment]
Hi @ericxiao251, yes those would be environment variables, however you could also configure them in the airflow.cfg

**Comment by ericxiao251** [comment]
@dintorf would you be able to help me debug my problem? I think somewhere in my setup I must be overriding the `BROKER_URL` or `AIRFLOW__CELERY__BROKER_URL` but I can't seem to pinpoint it.


I'm using Airflow 1.18.0 helm charts, I have this defined in my values:

```yaml
airflow:
  enableBuiltInSecretEnvVars:
    AIRFLOW__CORE__FERNET_KEY: false
    AIRFLOW__WEBSERVER__SECRET_KEY: false

    AIRFLOW__CELERY__BROKER_URL: true
    AIRFLOW__CELERY__RESULT_BACKEND: true

  secret:
    - envName: AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME
      secretName: airflow-2-env-secret
      secretKey: AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME
    - envName: AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS
      secretName: airflow-2-env-secret
      secretKey: AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS
    - envName: BROKER_URL
      secretName: airflow-2-env-secret
      secretKey: BROKER_URL

  config:
    celery:
      celery_config_options: celery_config_extend.MY_CELERY_CONFIG

  flower:
    extraVolumes:
      - name: celery-config
        configMap:
          name: airflow-2-celery-config
    
    extraVolumeMounts:
      - name: celery-config
        mountPath: /usr/local/airflow/config/celery_config_extend.py
        subPath: celery_config_extend.py
```

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: airflow-2-celery-config
  labels:
    app.kubernetes.io/name: airflow
    app.kubernetes.io/component: celery
data:
  celery_config_extend.py: |
    from airflow.config_templates.default_celery import DEFAULT_CELERY_CONFIG

    MY_CELERY_CONFIG = {
        **DEFAULT_CELERY_CONFIG,
        broker_url: '{{ $secrets.AIRFLOW__CELERY__BROKER_URL }}',
    }

    if "broker_transport_options" in MY_CELERY_CONFIG:
        MY_CELERY_CONFIG["broker_transport_options"]["sentinel_kwargs"] = { "password": '{{ $secrets._REDIS_PASSWORD }}' }
        MY_CELERY_CONFIG["broker_transport_options"]["master_name"] = "mymaster"
    else:
        MY_CELERY_CONFIG["broker_transport_options"] = {
            "sentinel_kwargs": { "password": '{{ $secrets._REDIS_PASSWORD }}' }
            "master_name": "mymaster"
        }

```

```yaml
-- airflow-2-env-secret
apiVersion: v1
data:
  AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__MASTER_NAME: ++++++++
  AIRFLOW__CELERY_BROKER_TRANSPORT_OPTIONS__SENTINEL_KWARGS: ++++++++
  AIRFLOW__CELERY__BROKER_URL: ++++++++
  AIRFLOW__CELERY__CELERY_CONFIG_OPTIONS: ++++++++
  BROKER_URL: ++++++++
  _REDIS_PASSWORD: ++++++++
  connection: ++++++++
  redis_connection: ++++++++
```

---

I get this exception in my flower pod:
```
                               kombu.exceptions.OperationalError: [Errno 111] Connection refused
                               [2025-12-12 22:51:43,933] {connection.py:630} WARNING - No hostname was supplied. Reverting to default 'localhost'
                               [2025-12-12 22:51:43,983] {connection.py:630} WARNING - No hostname was supplied. Reverting to default 'localhost'
```

I get this exception in my worker pod:
```
                               [2025-12-12 22:51:26,400: WARNING/MainProcess] No hostname was supplied. Reverting to default 'localhost'
                               
                                -------------- celery@airflow-2-worker-0 v5.2.7 (dawn-chorus)
                               --- ***** -----
                               -- ******* ---- Linux-6.1.158-x86_64-with-debian-11.3 2025-12-12 22:51:26
                               - *** --- * ---
                               - ** ---------- [config]
                               - ** ---------- .> app:         airflow.executors.celery_executor:0x7f2b9ba2c0d0
                               - ** ---------- .> transport:   amqp://guest:**@localhost:5672//
                               - ** ---------- .> results:     disabled://
                               - *** --- * --- .> concurrency: 16 (prefork)
                               -- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
                               --- ***** -----
                                -------------- [queues]
                                               .> default          exchange=default(direct) key=default
                               
                               
                               [tasks]
                                 . airflow.executors.celery_executor.execute_command
                               
2025-12-12T22:51:28.418646281Z [2025-12-12 22:51:28,418: ERROR/MainProcess] consumer: Cannot connect to amqp://guest:**@127.0.0.1:5672//: [Errno 111] Connection refused.
                               Trying again in 2.00 seconds... (1/100)
```
